R/table.R

```diff
@@ -0,0 +1,105 @@
+#' Generics for tabular vectors
+#'
+#' @description
+#'
+#' \Sexpr[results=rd, stage=render]{vctrs:::lifecycle("experimental")}
+#'
+#' These generics make it possible to work with tables, i.e. vectors
+#' of dimension 2 like data frames and matrices. The main motivation
+#' of this API is to support colwise operations such as slicing
+#' columns. Operating across columns requires slightly different
+#' semantics and concepts than rowwise operations:
+#'
+#' - A tabular prototype is a zero-rows, zero-cols table. In order for
+#'   your tabular class to be compatible with vctrs, it needs to
+#'   support such empty instances.
+#'
+#' - The vector type of a data frame includes its columns (their names
+#'   and their types). The tabular type of a data frame doesn't. Hence
+#'   `tbl_ptype2()` and `tbl_cast()` accept inputs of any shape,
+#'   whereas `vec_ptype2()` and `vec_cast()` require congruent shapes
+#'   and names.
+#'
+#' The details and theory of tabular operations are in development and
+#' likely to change in the future.
+#'
+#' @param x A tabular vector.
+#'
+#' @export
+tbl_is <- function(x) {
+  .Call(vctrs_tbl_is, x)
+}
+#' @rdname tbl_is
+#' @inheritParams vec_assert
+#' @export
+tbl_assert <- function(x, arg = as_label(substitute(x))) {
+  .Call(vctrs_tbl_assert, x, arg)
+}
+
+#' @rdname tbl_is
+#' @export
+tbl_size <- function(x) {
+  .Call(vctrs_tbl_size, x)
+}
+
+#' @rdname tbl_is
+#' @inheritParams vec_slice
```

Comment:
Why do we need a separate `tbl_slice()`?

Reply:
To slice across columns with vctrs semantics (for instance supporting proxies with virtual columns). This will be needed in tibble to implement `[`.

Reply:
But isn't this already supported by `vec_slice()`?

Reply:
`vec_slice()` slices across rows, `rct_slice()` slices across columns. If you coerce the input to list and use `vec_slice()` to slice across columns, you lose the proxy semantics of the input class.